### PR TITLE
Avoid reset BlockDocsEnum#freqBuffer when indexHasFreq is false

### PR DESCRIPTION
**Pull request by easyice**
As discussion in https://github.com/apache/lucene/pull/12954, this is the second implementation can be considered. this change yielded a 13% speedup for `PKLookup`.

### CODE REVIEW

File: lucene/core/src/java/org/apache/lucene/codecs/lucene99/Lucene99PostingsReader.java [hunk 1]
```diff
@@ -399,7 +400,9 @@ public PostingsEnum reset(IntBlockTermState termState, int flags) throws IOExcep
       doc = -1;
       this.needsFreq = PostingsEnum.featureRequested(flags, PostingsEnum.FREQS);
       this.isFreqsRead = true;
-      if (indexHasFreq == false || needsFreq == false) {
+      // if reuse, we don't consider whether the needsFreq is changed. this is undefined in freq().
+      // see also: BasePostingsFormatTestCase#testPostingsEnumFreqs
+      if ((indexHasFreq == false || needsFreq == false) && reuse == false) {
```
[the remark this answers is not in the document]
Comment (easyice) [reply lucene/core/src/java/org/apache/lucene/codecs/lucene99/Lucene99PostingsReader.java:405]: Thank you for explaining @jpountz! makes sense to me, the only problem is the `PostingsEnum#freq()` method might return 0, although this is undefined, is this looks good?

**Comment by jpountz** [comment]
Sorry I might have put you on the wrong track, I didn't know we had such leniency around frequencies/positions/offsets. Out of curiosity, I tried to add checks to AssertingLeafReader to fail when reading freqs/positions/offsets/payloads if they have not been requested in the flags, and there are many test failures. It's not entirely clear to me how intentional that is, @rmuir do you have context by any chance?

For the case you're trying to improve, another approach could consist of filling at most `docFreq` entries in the freq buffer:

```patch
diff --git a/lucene/core/src/java/org/apache/lucene/codecs/lucene99/Lucene99PostingsReader.java b/lucene/core/src/java/org/apache/lucene/codecs/lucene99/Lucene99PostingsReader.java
index 80d24917477..362fb34539e 100644
--- a/lucene/core/src/java/org/apache/lucene/codecs/lucene99/Lucene99PostingsReader.java
+++ b/lucene/core/src/java/org/apache/lucene/codecs/lucene99/Lucene99PostingsReader.java
@@ -400,9 +400,9 @@ public final class Lucene99PostingsReader extends PostingsReaderBase {
       this.needsFreq = PostingsEnum.featureRequested(flags, PostingsEnum.FREQS);
       this.isFreqsRead = true;
       if (indexHasFreq == false || needsFreq == false) {
-        for (int i = 0; i < ForUtil.BLOCK_SIZE; ++i) {
-          freqBuffer[i] = 1;
-        }
+        // Filling this buffer may not be cheap when doing primary key lookups, so we make sure to
+        // not fill more than `docFreq` entries.
+        Arrays.fill(freqBuffer, 0, Math.min(ForUtil.BLOCK_SIZE, docFreq), 1);
       }
       accum = 0;
       blockUpto = 0;

```

**Comment by easyice** [comment]
> another approach could consist of filling at most docFreq entries in the freq buffer

It's a nice idea!

> I didn't know we had such leniency around frequencies/positions/offsets.

Yes, It feels like it should be more clear.


I found there is a big different speedup between java versions, the benchmark shows a significant speedup on java17, but only 3% on java21

**java17:**
```
    TaskQPS baseline      StdDevQPS my_modified_version      StdDev                Pct diff p-value
PKLookup       89.14     (10.0%)      104.32      (9.1%)   17.0% (  -1% -   40%) 0.000               
```

**java21:**
```
    TaskQPS baseline      StdDevQPS my_modified_version      StdDev                Pct diff p-value
PKLookup      113.20     (11.9%)      116.58     (10.7%)    3.0% ( -17% -   28%) 0.402
```

**Comment by rmuir** [comment]
> I tried to add checks to AssertingLeafReader to fail when reading freqs/positions/offsets/payloads if they have not been requested in the flags, and there are many test failures. It's not entirely clear to me how intentional that is, @rmuir do you have context by any chance?

@jpountz The current behavior is documented: https://github.com/apache/lucene/blob/main/lucene/core/src/java/org/apache/lucene/index/PostingsEnum.java#L73-L99

**Comment by rmuir** [comment]
as far as failing the methods, i'm not really sure if it is a good idea for frequencies. Treating the value as 1 seems reasonable, and e.g. i'm pretty sure TermScorer works correctly because of it, the `omitTF` has behaved like this for a really long time as far as i know, as the typical use case for this was super-short documents where frequency doesn't matter.

**Pull request merged**